src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp

```diff
@@ -21,20 +21,48 @@ namespace ann {
 //! BackwardVisitor visitor class.
 inline BackwardVisitor::BackwardVisitor(arma::mat&& input,
                                         arma::mat&& error,
-                                        arma::mat&& delta) :
+                                        arma::mat&& delta,
+                                        size_t layer) :
   input(std::move(input)),
   error(std::move(error)),
-  delta(std::move(delta))
+  delta(std::move(delta)),
+  index(layer)
 {
   /* Nothing to do here. */
 }
 
 template<typename LayerType>
 inline void BackwardVisitor::operator()(LayerType* layer) const
+{
+  LayerBackward(layer, layer->OutputParameter());
+}
+
+template<typename T>
+inline typename std::enable_if<
+    !HasRunCheck<T, bool&(T::*)(void)>::value, void>::type
+BackwardVisitor::LayerBackward(T* layer, arma::mat& /* input */) const
 {
   layer->Backward(std::move(input), std::move(error), std::move(delta));
 }
 
+template<typename T>
+inline typename std::enable_if<
+    HasRunCheck<T, bool&(T::*)(void)>::value, void>::type
+BackwardVisitor::LayerBackward(T* layer, arma::mat& /* input */) const
+{
+  if (index  == -1)
```

Comment:
We compare against an unsigned integer (``size_t``) so this is never true.

Reply:
i have changed to int now